primitives/src/persistent.rs

```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    impl<T: Eq + Hash> FromIterator<T> for PHashTrieSet<T> {
+        fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+            Self(rpds::HashTrieSetSync::from_iter(iter))
+        }
+    }
+
+    impl<T: TypeId + Eq + Hash> TypeId for PHashTrieSet<T> {
+        fn id() -> String {
+            format!("Vec<{}>", T::id())
+        }
+    }
+
+    impl<T: IntoSchema + Eq + Hash> IntoSchema for PHashTrieSet<T> {
+        fn type_name() -> String {
+            format!("Vec<{}>", T::type_name())
+        }
+        fn update_schema_map(map: &mut iroha_schema::MetaMap) {
+            if !map.contains_key::<Self>() {
+                map.insert::<Self>(iroha_schema::Metadata::Vec(iroha_schema::VecMeta {
+                    ty: core::any::TypeId::of::<T>(),
+                }));
+
+                T::update_schema_map(map);
+            }
+        }
+    }
+}
+
+/// Persistent hash map
+pub mod map {
+    use super::*;
+
+    /// Persistent map
+    #[derive(Serialize, Deserialize, Deref, DerefMut, Clone, Debug)]
+    pub struct PHashTrieMap<K: Eq + Hash, V>(rpds::HashTrieMapSync<K, V>);
+
+    impl<K: Eq + Hash, V> PHashTrieMap<K, V> {
+        /// Construct [`Self`]
+        pub fn new() -> Self {
+            Self(rpds::HashTrieMap::new_sync())
+        }
+    }
+
+    impl<K: Eq + Hash, V> Default for PHashTrieMap<K, V> {
+        fn default() -> Self {
+            Self(rpds::HashTrieMapSync::default())
+        }
+    }
+
+    impl<K: Encode + Eq + Hash, V: Encode> Encode for PHashTrieMap<K, V> {
+        #[allow(clippy::arithmetic_side_effects)]
+        fn size_hint(&self) -> usize {
+            mem::size_of::<u32>() + (mem::size_of::<K>() + mem::size_of::<V>()) * self.size()
+        }
+
+        fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+            let len = self.size();
+            Compact(u32::try_from(len).expect("Length is above u32::max")).encode_to(dest);
+
+            for element in self.iter() {
+                element.encode_to(dest);
+            }
+        }
+    }
+
+    impl<K: Decode + Eq + Hash, V: Decode> Decode for PHashTrieMap<K, V> {
+        fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+            <Compact<u32>>::decode(input).and_then(move |Compact(len)| {
+                input.descend_ref()?;
+                let result = (0..len).map(|_| Decode::decode(input)).collect();
+                input.ascend_ref();
+                result
+            })
+        }
+    }
+
+    impl<K: Eq + Hash, V> FromIterator<(K, V)> for PHashTrieMap<K, V> {
+        fn from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
+            Self(rpds::HashTrieMapSync::from_iter(iter))
+        }
+    }
+
+    impl<K: TypeId + Eq + Hash, V: TypeId> TypeId for PHashTrieMap<K, V> {
+        fn id() -> String {
+            format!("Map<{}, {}>", K::id(), V::id())
+        }
+    }
+
+    impl<K: IntoSchema + Eq + Hash, V: IntoSchema> IntoSchema for PHashTrieMap<K, V> {
```

Comment:
Same concert about containing structs with hashes inside our schema

Reply:
Replied in previous comment.